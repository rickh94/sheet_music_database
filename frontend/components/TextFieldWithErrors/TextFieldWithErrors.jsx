import React from 'react'
import PropTypes from 'prop-types'
import {
  Field,
  Label,
  Input,
  Help,
  Control
} from 'react-bulma-components/lib/components/form'

function TextFieldWithErrors(props) {
  return (
    <Field className={props.className}>
      {props.label && <Label>{props.label}</Label>}
      <Control className="is-expanded">
        <Input
          type={props.type}
          placeholder={props.placeholder}
          value={props.value}
          onChange={e => props.onChange(e, props.name)}
          color={props.error ? 'danger' : null}
          className={`is-expanded ${props.inputClass}`}
        />
      </Control>
      {props.error && <Help color="danger">{props.error}</Help>}
    </Field>
  )
}

TextFieldWithErrors.propTypes = {
  label: PropTypes.string,
  type: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  error: PropTypes.any,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  group: PropTypes.bool,
  className: PropTypes.string,
  inputClass: PropTypes.string
}

export default TextFieldWithErrors
