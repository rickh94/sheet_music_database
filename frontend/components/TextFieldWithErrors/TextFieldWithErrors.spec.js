import React from 'react'
import { shallow } from 'enzyme'
import TextFieldWithErrors from './TextFieldWithErrors'

import '../../setupTests'

describe('<TextFieldWithErrors />', () => {
  const wrapper = shallow(
    <TextFieldWithErrors
      label="Test"
      type="text"
      value="test"
      onChange={jest.fn()}
      error={null}
      name="test"
    />
  )

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a control', () => {
    expect(wrapper.exists('Control')).toBeTruthy()
  })

  it('renders an input', () => {
    expect(wrapper.exists('Label')).toBeTruthy()
    expect(wrapper.find('Label').contains('Test'))
  })

  it('renders a Field', () => {
    expect(wrapper.exists('Field')).toBeTruthy()
  })

  it('renders errors', () => {
    const error = 'test error'
    wrapper.setProps({ error })
    expect(wrapper.exists('Help')).toBeTruthy()
    expect(wrapper.find('Help').contains(error)).toBeTruthy()
  })

  it('calls onChange callback when a change occurs', () => {
    const onChange = jest.fn()
    wrapper.setProps({ onChange })
    wrapper.find('Input').simulate('change', { target: 'testvalue' })
    expect(onChange).toHaveBeenCalled()
  })

  it('does not render label if not provided', () => {
    wrapper.setProps({label: null})
    expect(wrapper.exists('Label')).toBeFalsy()
  })
})
