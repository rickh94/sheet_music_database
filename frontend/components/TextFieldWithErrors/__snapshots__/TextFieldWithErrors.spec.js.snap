// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextFieldWithErrors /> renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextFieldWithErrors
    error={null}
    label="Test"
    name="test"
    onChange={[MockFunction]}
    type="text"
    value="test"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "align": null,
      "backgroundColor": undefined,
      "children": Array [
        <Label
          className=""
          htmlFor={null}
          size={null}
          style={Object {}}
        >
          Test
        </Label>,
        <Control
          className=""
          fullwidth={false}
          iconLeft={false}
          iconRight={false}
          loading={false}
          renderAs="div"
          size={null}
          style={Object {}}
        >
          <Input
            className=""
            color={null}
            disabled={false}
            isStatic={false}
            name={null}
            onChange={[Function]}
            placeholder=""
            readOnly={false}
            size={null}
            style={Object {}}
            type="text"
            value="test"
          />
        </Control>,
        null,
      ],
      "className": "",
      "clearfix": undefined,
      "clipped": undefined,
      "hidden": undefined,
      "horizontal": false,
      "invisible": undefined,
      "italic": undefined,
      "kind": null,
      "marginless": undefined,
      "multiline": false,
      "overlay": undefined,
      "paddingless": undefined,
      "pull": undefined,
      "radiusless": undefined,
      "renderAs": "div",
      "responsive": undefined,
      "shadowless": undefined,
      "style": Object {},
      "textAlignment": undefined,
      "textColor": undefined,
      "textSize": undefined,
      "textTransform": undefined,
      "textWeight": undefined,
      "unselectable": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "backgroundColor": undefined,
          "children": "Test",
          "className": "",
          "clearfix": undefined,
          "clipped": undefined,
          "hidden": undefined,
          "htmlFor": null,
          "invisible": undefined,
          "italic": undefined,
          "marginless": undefined,
          "overlay": undefined,
          "paddingless": undefined,
          "pull": undefined,
          "radiusless": undefined,
          "responsive": undefined,
          "shadowless": undefined,
          "size": null,
          "style": Object {},
          "textAlignment": undefined,
          "textColor": undefined,
          "textSize": undefined,
          "textTransform": undefined,
          "textWeight": undefined,
          "unselectable": undefined,
        },
        "ref": null,
        "rendered": "Test",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "backgroundColor": undefined,
          "children": <Input
            className=""
            color={null}
            disabled={false}
            isStatic={false}
            name={null}
            onChange={[Function]}
            placeholder=""
            readOnly={false}
            size={null}
            style={Object {}}
            type="text"
            value="test"
          />,
          "className": "",
          "clearfix": undefined,
          "clipped": undefined,
          "fullwidth": false,
          "hidden": undefined,
          "iconLeft": false,
          "iconRight": false,
          "invisible": undefined,
          "italic": undefined,
          "loading": false,
          "marginless": undefined,
          "overlay": undefined,
          "paddingless": undefined,
          "pull": undefined,
          "radiusless": undefined,
          "renderAs": "div",
          "responsive": undefined,
          "shadowless": undefined,
          "size": null,
          "style": Object {},
          "textAlignment": undefined,
          "textColor": undefined,
          "textSize": undefined,
          "textTransform": undefined,
          "textWeight": undefined,
          "unselectable": undefined,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "backgroundColor": undefined,
            "className": "",
            "clearfix": undefined,
            "clipped": undefined,
            "color": null,
            "disabled": false,
            "hidden": undefined,
            "invisible": undefined,
            "isStatic": false,
            "italic": undefined,
            "marginless": undefined,
            "name": null,
            "onChange": [Function],
            "overlay": undefined,
            "paddingless": undefined,
            "placeholder": "",
            "pull": undefined,
            "radiusless": undefined,
            "readOnly": false,
            "responsive": undefined,
            "shadowless": undefined,
            "size": null,
            "style": Object {},
            "textAlignment": undefined,
            "textColor": undefined,
            "textSize": undefined,
            "textTransform": undefined,
            "textWeight": undefined,
            "type": "text",
            "unselectable": undefined,
            "value": "test",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      null,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "align": null,
        "backgroundColor": undefined,
        "children": Array [
          <Label
            className=""
            htmlFor={null}
            size={null}
            style={Object {}}
          >
            Test
          </Label>,
          <Control
            className=""
            fullwidth={false}
            iconLeft={false}
            iconRight={false}
            loading={false}
            renderAs="div"
            size={null}
            style={Object {}}
          >
            <Input
              className=""
              color={null}
              disabled={false}
              isStatic={false}
              name={null}
              onChange={[Function]}
              placeholder=""
              readOnly={false}
              size={null}
              style={Object {}}
              type="text"
              value="test"
            />
          </Control>,
          null,
        ],
        "className": "",
        "clearfix": undefined,
        "clipped": undefined,
        "hidden": undefined,
        "horizontal": false,
        "invisible": undefined,
        "italic": undefined,
        "kind": null,
        "marginless": undefined,
        "multiline": false,
        "overlay": undefined,
        "paddingless": undefined,
        "pull": undefined,
        "radiusless": undefined,
        "renderAs": "div",
        "responsive": undefined,
        "shadowless": undefined,
        "style": Object {},
        "textAlignment": undefined,
        "textColor": undefined,
        "textSize": undefined,
        "textTransform": undefined,
        "textWeight": undefined,
        "unselectable": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "backgroundColor": undefined,
            "children": "Test",
            "className": "",
            "clearfix": undefined,
            "clipped": undefined,
            "hidden": undefined,
            "htmlFor": null,
            "invisible": undefined,
            "italic": undefined,
            "marginless": undefined,
            "overlay": undefined,
            "paddingless": undefined,
            "pull": undefined,
            "radiusless": undefined,
            "responsive": undefined,
            "shadowless": undefined,
            "size": null,
            "style": Object {},
            "textAlignment": undefined,
            "textColor": undefined,
            "textSize": undefined,
            "textTransform": undefined,
            "textWeight": undefined,
            "unselectable": undefined,
          },
          "ref": null,
          "rendered": "Test",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "backgroundColor": undefined,
            "children": <Input
              className=""
              color={null}
              disabled={false}
              isStatic={false}
              name={null}
              onChange={[Function]}
              placeholder=""
              readOnly={false}
              size={null}
              style={Object {}}
              type="text"
              value="test"
            />,
            "className": "",
            "clearfix": undefined,
            "clipped": undefined,
            "fullwidth": false,
            "hidden": undefined,
            "iconLeft": false,
            "iconRight": false,
            "invisible": undefined,
            "italic": undefined,
            "loading": false,
            "marginless": undefined,
            "overlay": undefined,
            "paddingless": undefined,
            "pull": undefined,
            "radiusless": undefined,
            "renderAs": "div",
            "responsive": undefined,
            "shadowless": undefined,
            "size": null,
            "style": Object {},
            "textAlignment": undefined,
            "textColor": undefined,
            "textSize": undefined,
            "textTransform": undefined,
            "textWeight": undefined,
            "unselectable": undefined,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "backgroundColor": undefined,
              "className": "",
              "clearfix": undefined,
              "clipped": undefined,
              "color": null,
              "disabled": false,
              "hidden": undefined,
              "invisible": undefined,
              "isStatic": false,
              "italic": undefined,
              "marginless": undefined,
              "name": null,
              "onChange": [Function],
              "overlay": undefined,
              "paddingless": undefined,
              "placeholder": "",
              "pull": undefined,
              "radiusless": undefined,
              "readOnly": false,
              "responsive": undefined,
              "shadowless": undefined,
              "size": null,
              "style": Object {},
              "textAlignment": undefined,
              "textColor": undefined,
              "textSize": undefined,
              "textTransform": undefined,
              "textWeight": undefined,
              "type": "text",
              "unselectable": undefined,
              "value": "test",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        null,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
